internal/gensample/gensample.go

```diff
@@ -0,0 +1,757 @@
+	p("  return err")
+	p("}")
+	p("")
+
+	return g.handleOut(meth, respConfs, nil)
+}
+
+func (g *generator) paging(meth *descriptor.MethodDescriptorProto, pf *descriptor.FieldDescriptorProto, respConfs []schema_v1p2.ResponseConfig) error {
+	p := g.pt.Printf
+
+	p("it := c.%s(ctx, req)", meth.GetName())
+	p("for {")
+	p("  resp, err := it.Next()")
+	p("  if err == iterator.Done {")
+	p("    break")
+	p("  }")
+	p("  if err != nil {")
+	p("    return err")
+	p("  }")
+
+	var typ initType
+	if tn := pf.GetTypeName(); tn != "" {
+		typ = initType{desc: g.descInfo.Type[tn]}
+	} else {
+		typ = initType{prim: pf.GetType()}
+	}
+
+	err := g.handleOut(meth, respConfs, &typ)
+
+	p("}")
+	p("")
+	g.imports[pbinfo.ImportSpec{Path: "google.golang.org/api/iterator"}] = true
+
+	return err
+}
+
+func (g *generator) lro(meth *descriptor.MethodDescriptorProto, methConf GAPICMethod, respConfs []schema_v1p2.ResponseConfig) error {
+	p := g.pt.Printf
+
+	p("op, err := c.%s(ctx, req)", meth.GetName())
+	p("if err != nil {")
+	p("  return err")
+	p("}")
+	p("")
+	p("resp, err := op.Wait(ctx)")
+	p("if err != nil {")
+	p("  return err")
+	p("}")
+	p("")
+
+	retType := methConf.LongRunning.ReturnType
+	if retType == "" {
+		return errors.E(nil, "LRO return type not given")
+	}
+
+	typ := initType{desc: g.descInfo.Type["."+retType]}
+	return g.handleOut(meth, respConfs, &typ)
+}
+
+func (g *generator) handleOut(meth *descriptor.MethodDescriptorProto, respConfs []schema_v1p2.ResponseConfig, respTyp *initType) error {
+	st := newSymTab(nil)
+	if respTyp != nil {
+		st.universe["$resp"] = true
+		st.scope["$resp"] = *respTyp
+	}
+
+	for _, out := range respConfs {
+		if err := writeOutputSpec(out, st, g); err != nil {
+			return errors.E(err, "cannot write output handling code")
+		}
+	}
+
+	if respTyp != nil && len(respConfs) == 0 {
+		g.pt.Printf("fmt.Println(resp)")
+		g.imports[pbinfo.ImportSpec{Path: "fmt"}] = true
+	}
+	return nil
+}
+
+// readSampleConfig loads sample configs from a file.
+func readSampleConfig(decoder *yaml.Decoder, fname string) (schema_v1p2.SampleConfig, error) {
+	var config schema_v1p2.SampleConfig
+	for true {
+		var sc schema_v1p2.SampleConfig
+		err := decoder.Decode(&sc)
+		if err != nil && err.Error() == "EOF" {
+			// last YAML document, all done
+			break
+		}
+		if err != nil {
+			return config, err
+		}
+		if sc.Type != expectedSampleConfigType {
+			// ignore non sample config YAMLs
+			continue
+		}
+		if sc.Version != expectedSampleConfigVersion {
+			// ignore unsupported versions
+			continue
+		}
+		config.Samples = append(config.Samples, sc.Samples...)
+	}
+	if len(config.Samples) == 0 {
+		return config, errors.E(nil, "Found no valid sample config in %q", fname)
+	}
+	return config, nil
+}
+
+// prependLines adds prefix to every line in b. A line is defined as a possibly empty run
+// of non-newlines terminated by a newline character.
+// If b doesn't end with a newline, prependLines panics.
```

Comment:
Can we return an error? I might've mentioned it before but using `panic` in a "library" is an anti-pattern. The caller should be able to handle the error as they see fit. 

Also, why can't a newline be appended in the event that it doesn't end with one?

Reply:
This function is mostly implemented by pongad@ but my guess is:

> why it panics instead of returning an error:

This is a pure internal helper function. The input of the function is always constructed by other functions in this program, independent of customer inputs. And therefore any error produced by this function is an internal error (someone screwed up and had a bug), and users can do nothing about it.

> why can't a newline be appended in the event that it doesn't end with one?

There isn't such a use case. This function is only used to prepend `//` to convert some code into comments, or prepend whitespace for indentation. The chunk of code to be prepended has to end with a newline, because otherwise following code will continue at the end of the last line of the code prepended, which is always wrong. We could force this function to append a newline character if it did not see one at the end of the buffer, but that might potentially result in more confusion in some corner cases in the future.

Anyway, I'd be happy to convert the panic into returning an error, but automatically adding a newline is a bit unnecessary for now IMHO. WDYT?

Reply:
Thanks for the detailed breakdown!

> and users can do nothing about it.

I'd say this is a reason _not_ to panic. If the calling code (i.e. the plugin) panics, that's different. I would be quite frustrated as a user if doing a normal invocation the program just crashes, especially without much context. I'd look to the calling functions to wrap the error in the call's context for reporting. Also it is harder to test edge cases when the code crashes.

>  I'd be happy to convert the panic into returning an error

Doesn't have to happen in this PR, but yes lets refactor this to return an error

Reply:
Just used three times altogether, I'll fix them here :)